zebra-chain/src/primitives/viewing_key.rs

```diff
@@ -0,0 +1,119 @@
+//! Type definitions for viewing keys and their hashes.
+
+use std::fmt;
+
+use zcash_primitives::{sapling::SaplingIvk, zip32::DiversifiableFullViewingKey};
+
+use crate::parameters::Network;
+
+pub mod sapling;
+
+#[cfg(test)]
+mod tests;
+
+const KEY_HASH_BYTE_SIZE: usize = 32;
+
+#[derive(Debug, Clone)]
+/// A Zcash Sapling or Orchard viewing key
+// TODO: Add Orchard types and any other Sapling key types
+pub enum ViewingKey {
+    /// An incoming viewing key for Sapling
+    SaplingIvk(Box<SaplingIvk>),
+
+    /// A diversifiable full viewing key for Sapling
+    DiversifiableFullViewingKey(Box<DiversifiableFullViewingKey>),
+}
+
+impl ViewingKey {
+    /// Returns an encoded byte representation of the viewing key
+    pub fn to_bytes(&self) -> Vec<u8> {
+        match self {
+            Self::SaplingIvk(ivk) => ivk.to_repr().to_vec(),
+            Self::DiversifiableFullViewingKey(dfvk) => dfvk.to_bytes().to_vec(),
+        }
+    }
+
+    /// Parses an encoded viewing key and returns it as a [`ViewingKey`] type.
+    pub fn parse(
+        key: &str,
+        network: Network,
+    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync + 'static>> // TODO: Consider using an explicit error type
+    {
+        Self::parse_extended_full_viewing_key(key, network)
+            .map(Self::DiversifiableFullViewingKey)
+            // TODO: Call .or_else() with every other parse method, they won't do anything unless the HRP is correct
+            .map_err(|err| err.to_string().into())
+    }
+}
+
+impl From<DiversifiableFullViewingKey> for ViewingKey {
+    fn from(value: DiversifiableFullViewingKey) -> Self {
+        Self::DiversifiableFullViewingKey(Box::new(value))
+    }
+}
+
+#[derive(Debug, PartialOrd, Ord, PartialEq, Eq)]
+/// The hash of a viewing key for use as an identifier and for authorizing remote access
```

Comment:
###### Optional Suggestion

Let's swap the attribute and comment as per https://github.com/rust-lang/style-team/issues/72.

```suggestion
/// The hash of a viewing key for use as an identifier and for authorizing remote access
#[derive(Debug, PartialOrd, Ord, PartialEq, Eq)]
```